### PR TITLE
Add method for retrieving ol.Overlay by id

### PR DESCRIPTION
**Pull request by jonataswalker**
~~This PR follows the same guidelines of https://github.com/openlayers/ol3/pull/2087.~~

This PR adds:
- [x] ol.Overlay.id
- [x] ol.Overlay#getId
- [x] ol.Map#getOverlayById

The reason is pretty obvious: Get an overlay without a loop iteration.

### CODE REVIEW

**Comment by ahocevar** [comment]
This could be much simpler if you didn't make `id` an `ol.Object` property. Then you wouldn't have to handle id changes. An id shouldn't change anyway. Just make the id an instance property instead, and don't add a setter, but only make it configurable upon instantiation.

You'll also have to add the `id` to `olx.OverlayOptions` in `externs/olx.js`.

**Comment by jonataswalker** [comment]
@ahocevar *"only make it configurable upon instantiation"*? This is quite different of `ol.Feature` getId/setId, is that you want?

```javascript
var overlay = new ol.Overlay({
    element: target,
    id: 'foo',
    position: [0, 0]
});
```

**Comment by ahocevar** [comment]
@jonataswalker That's what I had in mind. Would that be sufficient for your use case?

**Comment by jonataswalker** [comment]
Indeed, much simpler, bye bye lots of tests. Funny that I always wish to set `ol.Feature` id on constructor.

File: src/ol/map.js [hunk 1]
```diff
@@ -466,7 +473,7 @@ ol.Map = function(options) {
        * @this {ol.Map}
        */
       function(overlay) {
-        overlay.setMap(this);
+        this.addOverlayInternal(overlay);
       }, this);
```
Comment (ahocevar) [inline src/ol/map.js]: Can be replaced with
```js
this.overlays_.forEach(this.addOverlayInternal_, this);
```

File: src/ol/map.js [hunk 2]
```diff
@@ -540,6 +547,20 @@ ol.Map.prototype.addOverlay = function(overlay) {
 
 
 /**
+ * This deals with map's overlay collection changes.
+ * @param {ol.Overlay} overlay Overlay.
+ * @protected
```
Comment (ahocevar) [inline src/ol/map.js]: Should be `@private`.

File: src/ol/map.js [hunk 3]
```diff
@@ -1246,6 +1281,20 @@ ol.Map.prototype.removeOverlay = function(overlay) {
...
+ */
+ol.Map.prototype.removeOverlayInternal = function(overlay) {
```
Comment (ahocevar) [inline src/ol/map.js]: This method is only called once, so you can inline the code in the `ol.CollectionEventType.REMOVE` listener.

File: src/ol/overlay.js [hunk 4]
```diff
@@ -188,6 +194,16 @@ ol.Overlay.prototype.getElement = function() {
 
 
 /**
+ * Get the feature identifier which is set on constructor.
```
Comment (ahocevar) [inline src/ol/overlay.js]: This is not a *feature* identifier, but an *overlay* identifier. Also, does this need to be `@api`?

**Comment by ahocevar** [comment]
Thanks @jonataswalker. I added a few more suggestions. When you're done addressing these, we can merge this.

**Comment by jonataswalker** [comment]
@ahocevar thanks for review and improvement. All your notes are addressed.

**Comment by ahocevar** [comment]
Thanks @jonataswalker, looks great now.

**Pull request merged**